File: src/core/divisors.ts
const memo = new Map<number, number[]>();

// Computes the divisors of a given number.
//
// @remarks
// Uses a simple algorithm of trial division and memoizes the results.
//
// @returns The divisors in ascending order.
export default function divisors(n: number): number[] {
  if (memo.has(n)) {
    return memo.get(n)!;
  }

  const set = new Set<number>();
  const limit = Math.floor(Math.sqrt(n));

  // We only need to go up to the square root of the number to find all of its divisors; divisors always occur in pairs,
  // and you can't get divisors larger than the square root.
  for (let i = 1; i <= limit; i++) {
    if (n % i === 0) {
      set.add(i);
      set.add(n / i);
    }
  }

  const ds = Array.from(set);
  ds.sort((a, b) => a - b);

  memo.set(n, ds);
  return ds;
}

File: project_euler/common/divisors.py
import functools
import math


@functools.lru_cache(maxsize=None)
def divisors(n: int) -> list[int]:
    ds: set[int] = set()

    # Use trial division to find all divisors of n, up to sqrt(n).  Only check up to sqrt(n) because you can't have a
    # divisors greater than sqrt(n) that isn't paired with a divisor less than sqrt(n).
    limit = int(math.sqrt(n)) + 1
    for i in range(1, limit):
        if n % i == 0:
            ds.add(i)
            ds.add(n // i)

    return sorted(ds)


def sigma(n: int) -> int:
    return sum(divisors(n))


def tau(n: int) -> int:
    return len(divisors(n))
